### PR TITLE
Exclude expired transactions when counting block size during block creation

### PR DESCRIPTION
**Pull request by vusirikala**
## Description
Exclude expired transactions when counting block size during block creation

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [x] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: consensus/consensus-types/src/common.rs [hunk 1]
```diff
@@ -51,6 +51,36 @@ impl fmt::Display for TransactionSummary {
     }
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Deserialize, Serialize, Hash, Ord, PartialOrd)]
+pub struct TransactionSynopsis {
```
Comment (sitalkedia) [inline consensus/consensus-types/src/common.rs:55]: May be call it `TxnSummaryWithExpiration`?

Comment (sitalkedia) [reply consensus/consensus-types/src/common.rs:55]: Actually instead of creating this new struct, does it make sense to add expiration time to the txn summary itself?

Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:55]: Txn summary is used in a bunch of other places. Didn't want to change all of those for the sake of this small PR.
I can change `TxnSynopsis` to `TxnSummaryWithExpiration`.

Comment (sitalkedia) [reply consensus/consensus-types/src/common.rs:55]: Hmm, I did take a look and it seems trivial to add `expiration_timestamp_secs` to the `TransactionSummary` itself.  What am I missing?

Comment (vusirikala) [reply consensus/consensus-types/src/common.rs]: There are 189 other occurrences of `TransactionSummary` in the code. Have to change how the transaction summary is calculated in each of those cases. Not sure if that is worth it.
Screenshot 2024-07-01 at 2 35 35 PM

Comment (sitalkedia) [reply consensus/consensus-types/src/common.rs]: You don't have to change all of the above places, only place you need to change is when you call the constructor - may be ~4 occurrences? 

Screenshot 2024-07-01 at 2 41 35 PM

Comment (vusirikala) [reply consensus/consensus-types/src/common.rs]: Yeah. You are right. 
I tried changing all those occurrences. As `TxnPtr` is an alias for `TransactionSummary`, tried changing those instantiations as well. Faced an issue when instantiating `TransactionSummary` from `OrderedQueueKey` where this expiration time is not present.

File: consensus/src/quorum_store/utils.rs [hunk 2]
```diff
@@ -412,7 +412,11 @@ impl ProofQueue {
                             cur_unique_txns
                                 + txn_summaries
                                     .iter()
-                                    .filter(|txn_summary| !filtered_txns.contains(txn_summary))
+                                    .filter(|txn_summary| {
+                                        !filtered_txns.contains(txn_summary)
+                                            && txn_summary.expiration_timestamp_secs
```
Comment (bchocho) [inline consensus/src/quorum_store/utils.rs:417]: Isn't `latest_block_timestamp` in usecs and `expiration_timestamp_secs` in secs?

**Review by bchocho — CHANGES_REQUESTED** [review]

File: consensus/src/quorum_store/utils.rs (hunk 2 above)
Comment (bchocho) [reply consensus/src/quorum_store/utils.rs:417]: Also, don't you need to apply the same filter below when actually updating cur_unique_txns? Better yet, we should only apply the filter once?

Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs:417]: Thanks a lot for noticing this. Are you sure the `expiration_timestamp_secs` is not counted from the UNIX timestamp? I was searching for the usage of `expiration_timestamp_secs` in the code and found some usages like [this]
(https://github.com/aptos-labs/aptos-core/blob/main/api/goldens/aptos_api__tests__transactions_test__test_post_bcs_format_transaction.json#L7) where the `expiration_timestamp_secs` has been instantiated with a very large value.

 If `expiration_timestamp_secs` is not from the UNIX timestamp, then when is it from?

Comment (bchocho) [reply consensus/src/quorum_store/utils.rs:417]: It's counted from the UNIX epoch time, but the transaction expiration time is in `secs` and the block timestamp is in `usecs` so we have to convert units.

The examples are probably just examples. Transaction emitter is using seconds here https://github.com/aptos-labs/aptos-core/blob/9c2f7d32eed87842d8c6229c8a4700e1acf67ae1/sdk/src/transaction_builder.rs#L254

If you're unsure, maybe you can print some debug logs and run forge. In general, we should run some tests / write unit tests to ensure the expiration is working correctly.

File: consensus/src/quorum_store/types.rs [hunk 3]
```diff
@@ -58,14 +58,15 @@ impl PersistedValue {
         &self.maybe_payload
     }
 
-    pub fn summary(&self) -> Vec<TransactionSummary> {
+    pub fn synopsis(&self) -> Vec<TxnSummaryWithExpiration> {
```
Comment (bchocho) [inline consensus/src/quorum_store/types.rs:61]: Why is this renamed to `synopsis`?

File: consensus/src/quorum_store/utils.rs [hunk 4]
```diff
@@ -494,7 +494,13 @@ impl ProofQueue {
                             cur_unique_txns
                                 + txn_summaries
                                     .iter()
-                                    .filter(|txn_summary| !filtered_txns.contains(txn_summary))
+                                    .filter(|txn_summary| {
+                                        !filtered_txns.contains(txn_summary)
+                                                // latest_block_timestamp is microseonds since UNIX epoch
+                                                // expiration_timestamp_secs is seconds since UNIX epoch
+                                            && (txn_summary.expiration_timestamp_secs * 1_000_000)
```
Comment (bchocho) [inline consensus/src/quorum_store/utils.rs:501]: Can you add a const to this file, similar to https://github.com/aptos-labs/aptos-core/blob/533f3c33aa0c581da11bf3a14276be8fd40e19c8/crates/aptos/src/node/mod.rs#L59 ?

**Review by bchocho — COMMENTED** [review]
Can we add a simple unit test that exercises the expiration working as intended?

File: consensus/src/quorum_store/utils.rs (hunk 4 above)
Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs]: DONE.

File: consensus/src/quorum_store/types.rs (hunk 3 above)
Comment (vusirikala) [reply consensus/src/quorum_store/types.rs]: I actually used TransactionSynopsis earlier. After Sital's comment, I renamed it to TxnSummaryWithExpiration and forgot to update this statement. Updated it now.

File: consensus/src/quorum_store/utils.rs [hunk 5]
```diff
@@ -494,7 +499,14 @@ impl ProofQueue {
                             cur_unique_txns
                                 + txn_summaries
                                     .iter()
-                                    .filter(|txn_summary| !filtered_txns.contains(txn_summary))
+                                    .filter(|txn_summary| {
+                                        !filtered_txns.contains(txn_summary)
+                                                // latest_block_timestamp is microseonds since UNIX epoch
+                                                // expiration_timestamp_secs is seconds since UNIX epoch
+                                                // giving a second buffer for expiration as the expiration time is rounded off to seconds 
+                                            && ((txn_summary.expiration_timestamp_secs + 1) * MICRO_SEC_PER_SEC)
```
Comment (bchocho) [inline consensus/src/quorum_store/utils.rs:507]: We should be careful here.
* The multiplication as-is could overflow. So we should instead do integer division.
* We should match the move spec for expiration.

The move spec appears to be
https://github.com/aptos-labs/aptos-core/blob/533f3c33aa0c581da11bf3a14276be8fd40e19c8/aptos-move/framework/aptos-framework/sources/transaction_validation.move#L83-L86

where now_seconds() is the current block timestamp truncated to seconds.

So it appears the correct condition is:
`(self.latest_block_timestamp + 1) / MICRO_SEC_PER_SEC < txn_summary.expiration_timestamp_secs`
because the current block timestamp must be monotonically increasing from the previous timestamp.

If we want a truly accurate expiration condition, we can pass in the current block's timestamp from proposal_generator. I'm not sure that's necessary, as the number of truly unexpired transactions <= unique txn size based on the txn summaries we have.

Comment (vusirikala) [reply consensus/src/quorum_store/utils.rs:507]: If you are intending to approximate `self.latest_block_timestamp/MICRO_SEC_PER_SEC` with the next integer value, did you mean `(self.latest_block_timestamp + MICRO_SEC_PER_SEC) / MICRO_SEC_PER_SEC < txn_summary.expiration_timestamp_secs`?

**Review by bchocho — APPROVED** [review]

**Review by igor-aptos — APPROVED** [review]

**Review by sitalkedia — APPROVED** [review]
LG
